dev/dram/device.c

```diff
@@ -93,76 +154,451 @@ void dram_cbt_exit(dram_dev_t *dram,
+
+        if (packet->packet.packet.time >= wrdata_en_info.dfi_ts_start)
+        {
+            fill_wrdata_en_packet(packet, &wrdata_en_info, cycles_per_packet);
+        }
+
+        time_offset = 1;
+    } while (wrdata_en_info.phase_remaining || wrdata_info.phase_remaining >= phases_per_packet);
+
+    // Send CK packet tDHtrain + phy_wr_en after last packet
+    // Add 1 to offset to ensure beyond tDHtrain + phy_wr_en
+    time_offset = ((dram->cfg->t_sh_train + dram->cfg->phy_wr_en) >> ratio) + 1;
+
+    create_ck_packet_sequence(buffer, time_offset);
+
+    // Generate CA data packets to send training sequence
+    time_offset = ((dram->cfg->t_vref_ca_long - dram->cfg->t_sh_train) >> ratio) + 1;
+
+    packet = create_ck_packet_sequence(buffer, time_offset);
+    if (packet == NULL)
+    {
+        return WDDR_ERROR;
+    }
+
+    // Command
+    packet->packet.packet.cs_p0 = CS_HIGH << cs;
+    packet->packet.packet.cs_p1 = CS_HIGH << cs;
+    packet->packet.packet.address_p0 = command_address;
+
+    create_ck_packet_sequence(buffer, 1);
+
+    create_ck_packet_sequence(buffer, 2);
+
+    return WDDR_SUCCESS;
+}
+
+wddr_return_t dram_prepare_wrfifo_sequence(dram_dev_t *dram,
+                                           dfi_tx_packet_buffer_t *buffer,
+                                           burst_length_t burst_length,
+                                           chipselect_t cs,
+                                           uint16_t wrdata_en_offset,
+                                           command_data_t *data)
+{
+    uint16_t initial_dram_offset = 0;
+    uint8_t time_offset = 1;
+    packet_item_t *packet;
+    command_t command = {0};
+    packet_group_info_t wrdata_en_info, wrdata_info;
+
+    // Calculate timestamp of first CA packet (time 0).
+    // Shifted by ratio to be measured in DRAM CLK cycles.
+    initial_dram_offset = (buffer->ts_last_packet + time_offset) << dram->cfg->ratio;
+
+    // Calculate wr latency required for wr packet from time 0
+    // time 0 is identified as first CA packet ts
+    // wrdata offset is measured in DRAM CLKs
+    uint8_t wrdata_en_dram_clk = initial_dram_offset + CA_CLK_LEN + dram->cfg->phy_wr_lat + wrdata_en_offset;
+    uint8_t wrdata_dram_clk = wrdata_en_dram_clk + dram->cfg->phy_wr_en;
+    uint8_t cycles_per_packet = 1 << dram->cfg->ratio;
+
+    create_packet_group_info(&wrdata_en_info, dram->cfg->ratio, wrdata_en_dram_clk, burst_length + 4);
+    create_packet_group_info(&wrdata_info, dram->cfg->ratio, wrdata_dram_clk, burst_length);
+
+    // Create WRFIFO command sequence.
+    create_wrfifo_command(&command, data, cs);
+    dram_create_address_packet_sequence(buffer, dram->cfg->ratio, &command, time_offset);
+
+    time_offset = wrdata_en_info.dfi_ts_start - buffer->ts_last_packet;
+
+    // If enough delay is required, insert CKE packet to keep clocks cycling.
+    if (time_offset > 1)
+    {
+        create_cke_packet_sequence(buffer, 1);
+        time_offset--;
+    }
```

Comment:
so this actually comes in handy because we don't know if the next packet will be only 1 cke away or multiple.  We could move this into the do while loop where we keep adding multiple cke packets until we reach writing the wrdata packets, but like you said, we would waste cycles creating those packets and writing them into the buffer.  Definitely possible to do, but this could be a bit more advantageous.

Reply:
Yes, I understand. It's advantageous from perspective of using minimum number of packets, which is what I said above. But it's more code so from code size it's not advantageous. Since we're using O2 optimization (optimize for speed), I bet the compiler is inserting the contents of cke function directly in the if statement. If you remove that if statement, then the code will be removed saving some space.

Reply:
I did this on the side and it saves around 50 bytes total if updated in both functions. So I think given that, let's lean on side of speed and keep it as is.